k8s.go

```diff
@@ -0,0 +1,643 @@
+}
+
+func (c *k8sClient) updatePrefix() error {
+	var paths []*svbgptable.Path
+	node, err := c.k8scli.Nodes().Get(c.node, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+	cidr := node.Spec.PodCIDR
+	path, err := c.server.makePath(cidr, false)
+	log.Debugf("Set prefix: %#v", path)
+	paths = append(paths, path)
+	if err = c.server.updatePrefixSet(paths); err != nil {
+		return err
+	}
+	if _, err := c.server.bgpServer.AddPath("", paths); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *k8sClient) initialNeighborConfigSetting() error {
+	bgpconfig, err := c.getBGPConfig()
+	if err != nil {
+		return err
+	}
+	c.lastBgpconfig = bgpconfig
+	neighborConfigs, err := c.getNeighborConfigs(c.lastBgpconfig)
+	if err != nil {
+		return err
+	}
+	for _, n := range neighborConfigs {
+		if err = c.server.bgpServer.AddNeighbor(n); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// getNeighborConfigs returns the complete list of BGP neighbor configuration
+// which the node should peer.
+func (c *k8sClient) getNeighborConfigs(bgpconfig map[string]string) ([]*svbgpconfig.Neighbor, error) {
+	var neighbors []*svbgpconfig.Neighbor
+	if mesh, ok := bgpconfig[GlobalNodeMesh]; ok == false {
+		return nil, errors.New("mesh data not found")
+	} else if mesh == `{"enabled":true}` {
+		log.Debug("ebnable mesh")
+		ns, err := c.server.getMeshNeighborConfigs()
+		if err != nil {
+			return nil, err
+		}
+		neighbors = append(neighbors, ns...)
+	}
+	// --- Global peers ---
+	if ns, err := c.server.getGlobalNeighborConfigs(); err != nil {
+		return nil, err
+	} else {
+		neighbors = append(neighbors, ns...)
+	}
+	// --- Node-specific peers ---
+	if ns, err := c.server.getNodeSpecificNeighborConfigs(); err != nil {
+		return nil, err
+	} else {
+		neighbors = append(neighbors, ns...)
+	}
+	log.Debugf("neighbors=%s", neighbors)
+	return neighbors, nil
+}
+
+// checkBGPConfig checks the difference from the last BGP config information.
+// But in a case of the following changing, return error.
+//   /calico/bgp/v1/host/$NODENAME or /calico/global/as_num
+func (c *k8sClient) checkBGPConfig() error {
+	curBgpconfig, err := c.getBGPConfig()
+	if err != nil {
+		return nil
+	}
+	log.Debugf("checkBGPConfig lastBgpconfig: %s", c.lastBgpconfig)
+	if reflect.DeepEqual(c.lastBgpconfig, curBgpconfig) {
+		return nil
+	}
+	act := CompareMap(c.lastBgpconfig, curBgpconfig)
+	log.Debugf("checkBGPConfig action list: %#v", act)
+	for _, key := range act.Add {
+		if err := c.updateBGPConfig(Act_add, key, curBgpconfig); err != nil {
+			return err
+		}
+	}
+	for _, key := range act.Upd {
+		if err := c.updateBGPConfig(Act_upd, key, curBgpconfig); err != nil {
+			return err
+		}
+	}
+	for _, key := range act.Del {
+		if err := c.updateBGPConfig(Act_del, key, c.lastBgpconfig); err != nil {
+			return err
+		}
+	}
+	c.lastBgpconfig = curBgpconfig
+	return nil
+}
+
+func (c *k8sClient) updateBGPConfig(action string, key string, bgpconfig map[string]string) error {
+
```

Comment:
You are copying 140 LOC from main.go.

Reply:
We may be able to replace them with a common function.
But I made this patch on the assumption that I will refactor later. I also tried not to affect existing code.